File: 2130-maximum-product-of-the-length-of-two-palindromic-subsequences/2130-maximum-product-of-the-length-of-two-palindromic-subsequences.py
class Solution:
    def maxProduct(self, s: str) -> int:
        palindromes = {}

        for bitmask in range(1, 2 ** len(s)):
            current = [s[i] for i in range(len(s)) if (bitmask & (1 << i)) > 0]
            if current == current[::-1]:
                palindromes[bitmask] = len(current)
        
        best = 0
        for bitmask1 in palindromes:
            for bitmask2 in palindromes:
                if bitmask1 & bitmask2 == 0:
                    best = max(palindromes[bitmask1] * palindromes[bitmask2], best)
        
        return best

File: 2130-maximum-product-of-the-length-of-two-palindromic-subsequences/2130-maximum-product-of-the-length-of-two-palindromic-subsequences.cpp
class Solution {
public:
    int maxProduct(string s) {
        unordered_map<int, int> palindromeToLength;

        for (int i = 1; i < 1 << s.size(); i++) {
            string candidate = "";

            for (int j = 0; j < s.size(); j++) {
                if (i & (1 << j)) {
                    candidate += s[j];
                }
            }
            
            bool isPalindrome = true;
            for (int j = 0; j < candidate.size() / 2; j++) {
                if (candidate[j] != candidate[candidate.size() - 1 - j]) {
                    isPalindrome = false;
                    break;
                }
            }

            if (isPalindrome) {
                palindromeToLength[i] = candidate.size();
            }
        }

        int best = 0;

        for (auto const &[palindrome1, length1] : palindromeToLength) {
            for (auto const &[palindrome2, length2] : palindromeToLength) {
                if (!(palindrome1 & palindrome2)) {
                    best = max(best, length1 * length2);
                }
            }
        }

        return best;
    }
};